File: go/sort/sort_test.go
package sort

import (
	"testing"

	"gotest.tools/v3/assert"
)

func TestBubbleSort(t *testing.T) {
	arr := RandomArray(1000, 0, 10000)
	assert.Equal(t, false, IsAscArray(arr))
	BubbleSort(arr, CompareAsc)
	assert.Equal(t, true, IsAscArray(arr))

	arr = RandomArray(1000, 0, 10000)
	assert.Equal(t, false, IsDescArray(arr))
	BubbleSort2(arr, CompareDesc)
	assert.Equal(t, true, IsDescArray(arr))
}

func TestSelectionSort(t *testing.T) {
	arr := RandomArray(1000, 0, 10000)
	assert.Equal(t, false, IsAscArray(arr))
	SelectionSort(arr)
	assert.Equal(t, true, IsAscArray(arr))
}

func TestInsertionSort(t *testing.T) {
	arr := RandomArray(1000, 0, 10000)
	assert.Equal(t, false, IsAscArray(arr))
	InsertionSort(arr)
	assert.Equal(t, true, IsAscArray(arr))
}

func TestInsertionSortDesc(t *testing.T) {
	arr := RandomArray(1000, 0, 10000)
	assert.Equal(t, false, IsDescArray(arr))
	InsertionSortDesc(arr)
	assert.Equal(t, true, IsDescArray(arr))
}

func TestQuickSort(t *testing.T) {
	arr := RandomArray(9000000, 0, 90000000)
	assert.Equal(t, false, IsAscArray(arr))
	QuickSort(arr, 0, len(arr)-1)
	assert.Equal(t, true, IsAscArray(arr))
}

func TestGetMedianOfThree(t *testing.T) {
	arr := []int{5, 3, 7, 2, 4, 6, 8, 1}
	median := getMedianOfThree(arr, 0, len(arr)-1)
	assert.Equal(t, 3, median)

	arr = []int{6, 7, 8}
	median = getMedianOfThree(arr, 0, len(arr)-1)
	assert.Equal(t, 1, median)

	arr = []int{1, 2, 3}
	median = getMedianOfThree(arr, 0, len(arr)-1)
	assert.Equal(t, 1, median)

	arr = []int{77, 88, 77}
	median = getMedianOfThree(arr, 0, len(arr)-1)
	assert.Equal(t, 0, median)
}

func TestQuickSort1(t *testing.T) {
	arr := RandomArray(9000000, 0, 90000000)
	assert.Equal(t, false, IsAscArray(arr))
	QuickSort1(arr, 0, len(arr)-1)
	assert.Equal(t, true, IsAscArray(arr))
}

func TestMergeSort(t *testing.T) {
	const size = 1000
	arr := RandomArray(size, 0, 10000)
	assert.Equal(t, false, IsAscArray(arr))
	MergeSort(arr, make([]int, size), 0, size-1)
	assert.Equal(t, true, IsAscArray(arr))
}

File: Java/src/test/java/sort/SortTest.java
package sort;

import static org.junit.jupiter.api.Assertions.assertEquals;

import org.junit.jupiter.api.Test;

import com.example.sort.InsertionSort;
import com.example.sort.MergeSort;
import com.example.sort.QuickSort;
import com.example.sort.SelectionSort;
import com.example.sort.SortUtil;

public class SortTest {

    @Test
    public void testSelectionSort() {
        int[] array = SortUtil.randomArray(200, 1, 10000);
        SelectionSort.sort(array);
        assertEquals(true, SortUtil.isAscArray(array));
    }

    @Test
    public void testInsertionSort() {
        int[] array = SortUtil.randomArray(200, 1, 10000);
        InsertionSort.sort(array);
        assertEquals(true, SortUtil.isAscArray(array));
    }

    @Test
    public void testInsertionSortDesc() {
        int[] array = SortUtil.randomArray(200, 1, 10000);
        InsertionSort.sortDesc(array);
        assertEquals(true, SortUtil.isDescArray(array));
    }

    @Test
    public void testQuickSort() {
        int[] array = SortUtil.randomArray(200, 1, 10000);
        QuickSort.quickSort(array, 0, array.length - 1);
        assertEquals(true, SortUtil.isAscArray(array));
    }

    @Test
    public void testQuickSort1() {
        int[] array = SortUtil.randomArray(200, 1, 10000);
        QuickSort.quickSort1(array, 0, array.length - 1);   
        assertEquals(true, SortUtil.isAscArray(array));
    }


    @Test
    public void testMergeSort() {
        int[] array = SortUtil.randomArray(2000, 1, 10000);
        int[] temp = new int[array.length];
        MergeSort.sort(array, temp, 0, array.length - 1);
        assertEquals(true, SortUtil.isAscArray(array));

    }
}
